Mathlib/Tactic/ProdAssoc.lean

```diff
@@ -0,0 +1,151 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+import Mathlib.Lean.Expr.Basic
+import Mathlib.Logic.Equiv.Basic
+
+/-!
+# Associativity of products
+
+This file constructs a term elaborator for "obvious" equivalences between iterated products.
+For example,
+```lean
+(prod_assoc% : (α × β) × (γ × δ) ≃ α × (β × γ) × δ)
+```
+gives the "obvious" equivalence between `(α × β) × (γ × δ)` and `α × (β × γ) × δ`.
+-/
+
+namespace Lean.Expr
+
+open Lean Meta
+
+/-- A helper type to keep track of universe levels and types in iterated produts. -/
+inductive ProdTree where
+  | type (tp : Expr) (l : Level)
+  | prod (fst snd : ProdTree) (lfst lsnd : Level)
+deriving Repr
+
+/-- The iterated product corresponding to a `ProdTree`. -/
+def ProdTree.getType : ProdTree → Expr
+  | type tp _ => tp
+  | prod fst snd u v => mkAppN (.const ``Prod [u,v]) #[fst.getType, snd.getType]
+
+/-- The number of types appearing in an iterated product encoded as a `ProdTree`. -/
+def ProdTree.size : ProdTree → Nat
+  | type _ _ => 1
+  | prod fst snd _ _ => fst.size + snd.size
+
+/-- The components of an interated product, presented as a `ProdTree`. -/
+def ProdTree.components : ProdTree → List Expr
+  | type tp _ => [tp]
+  | prod fst snd _ _ => fst.components ++ snd.components
+
+/-- Make a `ProdTree` out of an `Expr`. -/
+partial def mkProdTree (e : Expr) : MetaM ProdTree :=
+  match e.consumeMData with
+    | .app (.app (.const ``Prod [u,v]) X) Y => do
+        return .prod (← X.mkProdTree) (← Y.mkProdTree) u v
+    | X => do
+      let some u := (← whnfD <| ← inferType X).type? | throwError "Not a type{indentExpr X}"
+      return .type X u
+
+/-- Given `P : ProdTree` representing an iterated product and `e : Expr` which
+should correspond to a term of the iterated product, this will return
+a list, whose items correspond to the leaves of `P` (i.e. the types appearing in the product),
+where each item is the appropriate composition of `Prod.fst` and `Prod.snd` applied to `e`
+resulting in an element of the type corresponding to the leaf.
+
+For example, if `P` corresponds to `(X × Y) × Z` and `t : (X × Y) × Z`, then this
+should return `[t.fst.fst, t.fst.snd, t.snd]`.
+-/
+def ProdTree.unpack (t : Expr) : ProdTree → MetaM (List Expr)
+  | type _ _ => return [t]
+  | prod fst snd u v => do
+      let fst' ← fst.unpack <| mkAppN (.const ``Prod.fst [u,v]) #[fst.getType, snd.getType, t]
+      let snd' ← snd.unpack <| mkAppN (.const ``Prod.snd [u,v]) #[fst.getType, snd.getType, t]
+      return fst' ++ snd'
+
+/-- This function should act as the "reverse" of `ProdTree.unpack`, constructing
+a term of the iterated product out of a list of terms of the types appearing in the product. -/
+def ProdTree.pack (ts : List Expr) : ProdTree → MetaM Expr
```

Comment:
I believe you could make this take an `Subarray Expr`, which would save you needing to do these to-and-from-array conversions.

Reply:
Good idea! I'll implement this in a follow up PR.